### PR TITLE
Create global roles

### PR DESCRIPTION
**Pull request by rmweir**
**MERGE ORDER:**
merge types-> update rancher w/ types-> merge rancher
**Problem:**
Current global roles do not have desired set of permissions for default role

**Solution:**
Allow admin to create global roles.

**Dependencies:**
https://github.com/rancher/types/pull/1018

**Issue:**
https://github.com/rancher/rancher/issues/16216

### CODE REVIEW

**Comment by dramich** [comment]
The validation of requests against builtin roles should follow the same pattern as roletemplates as seen [here](https://github.com/rancher/rancher/blob/3b22f8501526690bf470e721e6a2d592ef7880ca/pkg/api/customization/roletemplate/roletemplate.go#L19:18).

File: tests/integration/suite/test_global_roles.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
+def test_builtin_default_can_be_edited(admin_mc, remove_resource):
```
Comment (dramich) [inline tests/integration/suite/test_global_roles.py]: This test needs to undo anything it does, a fixture might be appropriate to ensure this happens

File: tests/integration/suite/test_global_roles.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    """Asserts admins can only edit a builtin global role's newUserDefault
+    field"""
+    admin_client = admin_mc.client
+
+    gr = admin_client.by_id_global_role(id="admin")
+    assert gr.builtin is True
+    assert "remove" not in gr.links.keys()
```
Comment (dramich) [inline tests/integration/suite/test_global_roles.py]: Assert on newUserDefault

File: tests/integration/suite/test_global_roles.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+                                                   displayName="gr-test",
+                                                   description="asdf",
+                                                   rules=None,
+                                                   newUserDefault=not
+                                                   gr.newUserDefault,
+                                                   builtin=True)
+    assert new_gr.name == gr.name
+    assert new_gr.get("description") == gr.description
+    assert new_gr.rules is not None
+    assert new_gr.get("builtin") == gr.builtin is True
```
Comment (dramich) [inline tests/integration/suite/test_global_roles.py]: `assert new_gr.get("builtin") is True`

File: tests/integration/suite/test_global_roles.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+    # newUserDefault is the only field that should editable
+    # for a builtin role
+    assert new_gr.newUserDefault == (not gr.newUserDefault)
```
Comment (dramich) [inline tests/integration/suite/test_global_roles.py]: `assert new_gr.newUserDefault is True`

File: tests/integration/suite/test_global_roles.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+    gr = admin_client.by_id_global_role(id="admin")
+    assert gr.builtin is True
+    assert "remove" not in gr.links.keys()
+
+    new_gr = admin_client.update_by_id_global_role(id=gr.id,
+                                                   displayName="gr-test",
+                                                   description="asdf",
+                                                   rules=None,
+                                                   newUserDefault=not
```
Comment (dramich) [inline tests/integration/suite/test_global_roles.py]: Set this explicitly to True

File: pkg/api/store/globalrole/store.go [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+}
+
+func Wrap(s types.Store, grLister v3.GlobalRoleLister) types.Store {
+	return &store{
+		Store:    s,
+		grLister: grLister,
+	}
+}
+
+func (s *store) Create(apiContext *types.APIContext, schema *types.Schema, data map[string]interface{}) (map[string]interface{}, error) {
```
Comment (dramich) [inline pkg/api/store/globalrole/store.go]: Create isn't needed if the only check is something norman already does.

File: tests/integration/suite/test_rbac.py [hunk 7]
```diff
@@ -761,6 +761,72 @@ def try_create_ns_secret():
     )
 
 
+def test_only_admin_can_crud_global_roles(admin_mc, user_mc, remove_resource):
```
Comment (dramich) [inline tests/integration/suite/test_rbac.py]: These two tests feel like they below under tests/integration/suite/test_global_roles.py.

Comment (dramich) [reply tests/integration/suite/test_rbac.py]: Realizing we have similar things in here already, it looks like it's becoming the `common` or `utils` of tests

File: app/rolebuilder.go [hunk 8]
```diff
@@ -54,6 +54,7 @@ func (rb *roleBuilder) addRole(displayName, name string) *roleBuilder {
 	if rb.name == "" {
 		rb.name = name
 		rb.displayName = displayName
+		rb.builtin = true
```
Comment (dramich) [inline app/rolebuilder.go]: Lets update newRoleBuilder to default to true for builtin

File: app/rolebuilder.go [hunk 9]
```diff
@@ -63,6 +64,7 @@ func (rb *roleBuilder) addRole(displayName, name string) *roleBuilder {
 	rb.next = &roleBuilder{
```
Comment (dramich) [inline app/rolebuilder.go]: Use newRoleBuilder then set the fields on the returned role

File: app/rolebuilder.go [hunk 10]
```diff
@@ -257,10 +259,12 @@ func (rb *roleBuilder) reconcileGlobalRoles(mgmt *config.ManagementContext) erro
 			return false, nil, errors.Errorf("unexpected type comparing %v and %v", have, want)
 		}
 
-		equal := haveGR.DisplayName == wantGR.DisplayName && reflect.DeepEqual(haveGR.Rules, wantGR.Rules)
+		builtin := haveGR.Builtin == wantGR.Builtin
+		equal := haveGR.DisplayName == wantGR.DisplayName && reflect.DeepEqual(haveGR.Rules, wantGR.Rules) && builtin
```
Comment (alena1108) [inline app/rolebuilder.go]: nit: builtin can be placed as a first condition. This way if it's false, you save on other conditions' evaluation. DeepEquals in particular

File: pkg/api/customization/globalrole/validator.go [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+	"github.com/rancher/norman/types"
+	v3 "github.com/rancher/types/apis/management.cattle.io/v3"
+)
+
+type Wrapper struct {
+	GlobalRoleLister v3.GlobalRoleLister
+}
+
+func (w Wrapper) Validator(request *types.APIContext, schema *types.Schema, data map[string]interface{}) error {
+	if request.Method != http.MethodPut && request.Method != http.MethodPost {
```
Comment (alena1108) [inline pkg/api/customization/globalrole/validator.go]: can we split this line into 2 checks? As one of them repeats at line 19

File: app/rolebuilder.go [hunk 12]
```diff
@@ -134,12 +135,12 @@ func (r *ruleBuilder) nonResourceURLs(u ...string) *ruleBuilder {
 	return r
 }
 
-func (r *ruleBuilder) addRoleTemplate(diplsayName, name, context string, builtin, external, hidden, administrative bool) *roleBuilder {
-	return r.rb.addRoleTemplate(diplsayName, name, context, builtin, external, hidden, administrative)
+func (r *ruleBuilder) addRoleTemplate(displayName, name, context string, builtin, external, hidden, administrative bool) *roleBuilder {
```
Comment (alena1108) [inline app/rolebuilder.go]: builtin can be dropped from the method signature

File: pkg/api/customization/globalrole/validator.go (hunk 11 above)
Comment (dramich) [reply pkg/api/customization/globalrole/validator.go]: The check should just be `if request.Method != http.MethodPut`

**Comment by alena1108** [comment]
LGTM

**Pull request merged**